lib/packaging/artifactory.rb

```
@@ -230,6 +232,66 @@ def package_name(platform_data, platform_tag)
       raise fail_message
     end
 
+    # Promotes a build based on git SHA or tag
+    # Depending on if it's an RPM or Deb package promote accordingly
+    # 'promote' by copying the package(s) to the enterprise directory on artifactory
+    #
+    # @param pkg [String] the package name ex. puppet-agent
+    # @param git_ref [String] full git SHA or tag of the package(s) to be promoted
+    # @param pe_version [String] enterprise version promoting to (XX.YY)
+    # @param platform_tag [String] the platform tag of the artifact
+    #   ex. el-7-x86_64, ubuntu-18.04-amd64
+    def promote_package(pkg, git_ref, pe_version, platform_tag)
+      package_ref = sha_or_tag(ref)
+      platform, version, arch = Pkg::Platforms.parse_platform_tag(platform_tag)
+      search_name = "#{pkg}-#{package_ref}.#{platform}#{version}.#{arch}"
+      noarch_search_name = "#{pkg}-#{package_ref}.#{platform}#{version}.noarch"
+      # if ubuntu get codename
+      if platform == "ubuntu"
+        codename = Pkg::Platforms.codename_for_platform_version(platform, version)
+        search_name = "#{pkg}_#{package_ref}*#{codename}_#{arch}"
+        noarch_search_name = "#{pkg}_#{package_ref}*#{codename}_all"
+      end
+      # search for the artifact, if we can't find it, try searching for the noarch version
+      artifacts_to_promote = Artifactory::Resource::Artifact.search(name: search_name, :artifactory_uri => @artifactory_uri)
+      if artifacts_to_promote.empty?
+        artifacts_to_promote = Artifactory::Resource::Artifact.search(name: noarch_search_name, :artifactory_uri => @artifactory_uri)
+      end
```

should we have a second check for whether or not `artifacts_to_promote` is empty and print a message saying nothing was found to promote? Not sure if that should be a failure case

---

I'm thinking failing in that case would be good - we don't want to assume things promoted, since nothing failed, but then be missing packages.

---

Because this task is run by iterating through all of the master platforms I can either just put a message in here every time something isn't found to promote or I can keep track of a return/counter value or something in the enterprise-dist promote.rake and if nothing from all platforms was found I can have it fail there. Which would we prefer?

---

I think we'd still want to fail immediately if something is missing. It seems not ideal if we managed to promote for some platforms and not others?

---

When promoting will we always be promoting for all platforms? If so I'm good with doing this I just wasn't sure.

---

If we don't always promote for all platforms, we could end up with different versions of the same package in the same version of PE, which seems.. not great to me, especially if we were intending to promote everything and something secretly failed. ¯\_(ツ)_/¯

---

Not every package is promoted for every platform (razor, for example)